pipeline/metadata/schema.py

```diff
@@ -0,0 +1,299 @@
+"""Schema for various data types passed around in the pipeline and bigquery"""
+
+from __future__ import absolute_import
+from __future__ import annotations  # required to use class as a type inside the class
+
+import dataclasses
+from dataclasses import dataclass
+from typing import Optional, List, Dict, Any
+
+# pylint: disable=too-many-instance-attributes
+
+
+@dataclass
+class SatelliteAnswer():
+  """Class for keeping track of Satellite answer content"""
+  # Keys
+  ip: str
+  # Metadata
+  asnum: Optional[int] = None
+  asname: Optional[str] = None
+  http: Optional[str] = None
+  cert: Optional[str] = None
+  matches_control: Optional[str] = None
+
+  def add_tag_to_answer(self, tag: SatelliteAnswerMetadata) -> None:
+    """Add tag information to a Satellite answer"""
+    if tag.asnum is not None:
+      self.asnum = tag.asnum
+    if tag.asname is not None:
+      self.asname = tag.asname
+    if tag.http is not None:
+      self.http = tag.http
+    if tag.cert is not None:
+      self.cert = tag.cert
+    if tag.matches_control is not None:
+      self.matches_control = tag.matches_control
+
+
+@dataclass
+class SatelliteAnswerMetadata(SatelliteAnswer):
```

Comment:
Why don't you use composition instead of inheritance?
It makes moving things around a lot easier (set a field instead of many) and it's easier to understand.

Reply:
The main thing I'm trying to achieve here is that the dataclasses being passed around match the ultimate structure we want to write to bigquery, (which is how it currently works with the dicts) so that when we get to the point of [writing the rows](https://github.com/censoredplanet/censoredplanet-analysis/blob/3cf6ac1ef0865bd2ccd616f1861b3701c6ef395f/pipeline/beam_tables.py#L466) we can just call the native `dataclass.asdict` and write them without having to add a translation layer that knows how to map the composed structure into a flat row.

Reply:
The serialization code is something you would write once. And you are writing something equivalent anyway in your dict update-like functions, with the difference that your logic is spread throughout different places.

It's important to decouple the structured used in code from the output format that has to stay backwards-compatible. Using the output format internally prevents us from being able to make changes.

My recommendation is to use composition and have a clear translation step at the end. It will free yourself to keep changing the internal format without breaking compatibility, and the serialization step will work as runnable documentation of how the output is generated.